From f819bab7bfead0cea9032a9f62edcdf2e8d75b0a Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Wed, 18 Feb 2026 17:24:14 +0400
Subject: [PATCH] fix: `DEAD_MAXS_Z` not getting set immediately

Resulting in things like players absorbing more lightning gun shots
on death than they should have at low `sv_fps 1`.

Closes https://github.com/WofWca/quake3-better-gibs-mod/issues/12.

Bug reproduction steps:
1. `\devmap q3dm1; addbot crash; set bot_pause 1;`,
   `give quad damage; give speed`.
2. `\set sv_fps 1`.
3. `\set timescale 8` for better reproduce-ability.
4. Aim at the Crash's torso (not the feet).
5. Shoot.

Sometimes Crash would get gibbed. Which is not the case
in the original game.

The bug has been introduced recently, in
"fix: shotgun not gibbing unless aiming at feet".
---
 code/game/g_combat.c | 26 ++++++++------------------
 code/game/g_local.h  | 13 +++++++++++++
 code/game/g_weapon.c | 16 +++++++++++++++-
 3 files changed, 36 insertions(+), 19 deletions(-)

diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index 1e18a41f..1e824432 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -601,24 +601,14 @@ void player_die( gentity_t *self, gentity_t *inflictor, gentity_t *attacker, int
 
 	self->s.loopSound = 0;
 
-	// The below line has been commented out in
-	// https://github.com/ec-/baseq3a/pull/49.
-	// Executing this line causes a bug where the shotgun doesn't gib
-	// unless you aim at the feet.
-	// See https://github.com/ioquake/ioq3/issues/794.
-	//
-	// Not executing this line makes is so that the corpse
-	// doesn't get shorter immediately on death
-	// and instead can still take up other pellets
-	// from the same shotgun shot.
-	//
-	// The purpose and the effect of this line is not entirely clear.
-	// Maybe it's to transition the player hitbox
-	// into the "lying down dead" state, make it shorter.
-	// But this is already handled in `PM_CheckDuck`,
-	// so maybe it's just leftover code.
-	//
-	// self->r.maxs[2] = DEAD_MAXS_Z;
+	// `SetDeadHeight` makes the corpse shorter.
+	// Executing this line unconditionally would cause a bug
+	// where the shotgun doesn't gib unless you aim at the feet,
+	// because, since the body is shorter, it would stop getting hit
+	// by other pellets from the same shot.
+	if ( !ShouldPostponeDeath( meansOfDeath ) ) {
+		SetDeadHeight( self );
+	}
 
 	// don't allow respawn until the death anim is done
 	// g_forcerespawn may force spawning at some later time
diff --git a/code/game/g_local.h b/code/game/g_local.h
index 10ec8dbf..6814d193 100644
--- a/code/game/g_local.h
+++ b/code/game/g_local.h
@@ -490,6 +490,19 @@ const char *BuildShaderStateConfig( void );
 //
 qboolean CanDamage (gentity_t *targ, vec3_t origin);
 void G_Damage (gentity_t *targ, gentity_t *inflictor, gentity_t *attacker, vec3_t dir, vec3_t point, int damage, int dflags, int mod);
+// The shotgun does `G_Damage` multiple times, per each pellet.
+// Normally that would mean that if the target is at 1 HP,
+// only one pellet would hit them.
+// But that would mean that the shotgun cannot gib. We don't want that,
+// so we postpone some of the effects of `G_Damage` until after
+// all the pellets have done their thing, in `ShotgunPattern`.
+// See
+// - https://github.com/ioquake/ioq3/issues/794.
+// - https://github.com/ec-/baseq3a/pull/49.
+// - https://github.com/WofWca/quake3-better-gibs-mod/issues/12.
+// - Also `glcient_s.damage_knockback`.
+#define ShouldPostponeDeath( mod ) (mod == MOD_SHOTGUN)
+#define SetDeadHeight( ent ) {ent->r.maxs[2] = DEAD_MAXS_Z;}
 qboolean G_RadiusDamage (vec3_t origin, gentity_t *attacker, float damage, float radius, gentity_t *ignore, int mod);
 int G_InvulnerabilityEffect( gentity_t *targ, vec3_t dir, vec3_t point, vec3_t impactpoint, vec3_t bouncedir );
 void body_die( gentity_t *self, gentity_t *inflictor, gentity_t *attacker, int damage, int meansOfDeath );
diff --git a/code/game/g_weapon.c b/code/game/g_weapon.c
index ea052cf0..664888c4 100644
--- a/code/game/g_weapon.c
+++ b/code/game/g_weapon.c
@@ -314,7 +314,7 @@ qboolean ShotgunPellet( vec3_t start, vec3_t end, gentity_t *ent ) {
 			// through a corpse.
 			// This is mostly to compensate for the balance changes
 			// that are introduced by the removal of the `self->r.maxs[2] = -8;`
-			// line in `player_die`.
+			// (`SetDeadHeight`) line in `player_die`.
 			// But it's probably also sensible otherwise that corpses
 			// affect "more serious" gameplay less.
 			// See
@@ -345,6 +345,7 @@ void ShotgunPattern( vec3_t origin, vec3_t origin2, int seed, gentity_t *ent ) {
 	vec3_t		end;
 	vec3_t		localForward, localRight, localUp;
 	qboolean	hitClient = qfalse;
+	gentity_t	*ent2;
 
 	// derive the right and up vectors from the forward vector, because
 	// the client won't have any other information
@@ -364,6 +365,19 @@ void ShotgunPattern( vec3_t origin, vec3_t origin2, int seed, gentity_t *ent ) {
 			ent->client->accuracy_hits++;
 		}
 	}
+
+	// Do what has been postponed in `G_Damage` due to `ShouldPostponeDeath`.
+	// assert( ShouldPostponeDeath( MOD_SHOTGUN ) );
+	ent2 = &g_entities[0];
+	for (i = 0; i < level.num_entities; i++, ent2++) {
+		if ( !ent2->inuse ) {
+			continue;
+		}
+
+		if ( ent2->client && ent2->client->ps.pm_type == PM_DEAD ) {
+			SetDeadHeight( ent2 );
+		}
+	}
 }
 
 
-- 
2.54.0

